### DISCUSSION TITLE
Discrepancy in Pair Correlation Function when Reloading Variational States in Potential Aziz Model

### DISCUSSION THREAD

**Discussion by MrAAafa99** [netket/netket#1483] [Q&A] [answered]
Hello,
 I want to save the potential Aziz variational states and plot the pair correlation function by reloading them. However, when I plot the pair correlation function without reloading, it differs from when I reload it again.
here is my code when I write the Variational State in "test.mpack" with flax, I also see [this document](https://netket.readthedocs.io/en/latest/docs/varstate.html) 

```python
import math
N = 10
d = 0.3 # 1/Angstrom
rm = 2.9673 # Angstrom
L = N / (0.3 * rm)

hilb = nk.hilbert.Particle(
    N=N,
    L=L,
    pbc=True)

def minimum_distance(x, sdim):
   ##"""Computes distances between particles using mimimum image convention""" ##
 n_particles = x.shape[0] // sdim 
 x = x.reshape(-1, sdim)

 distances = (-x[jnp.newaxis, :, :] + x[:, jnp.newaxis, :])[
    jnp.triu_indices(n_particles, 1)
    ]
 distances = jnp.remainder(distances + L / 2.0, L) - L / 2.0

 return jnp.linalg.norm(distances, axis=1)
sigma=1/math.sqrt(2)
epsilon=2
def potential(x, sdim):
  eps = 7.846373
  A = 0.544850 * 10 ** 6
  alpha = 13.353384
  c6 = 1.37332412
  c8 = 0.4253785
  c10 = 0.178100
  D = 1.241314
  dis = minimum_distance(x, sdim)
  return jnp.sum(
      eps
      * (
      A * jnp.exp(-alpha * dis)
      - (c6 / dis ** 6 + c8 / dis ** 8 + c10 / dis ** 10)
      * jnp.where(dis < D, jnp.exp(-((D / dis - 1) ** 2)), 1.0)
      )
      )

ekin = nk.operator.KineticEnergy(hilb, mass=1.0)
pot = nk.operator.PotentialEnergy(hilb, lambda x: potential(x, 1))
ha = ekin + pot
import matplotlib.pyplot as plt
import numpy as np
import netket as nk
import flax

sab = nk.sampler.MetropolisGaussian(hilbert=hilb, sigma=0.05,n_chains=16, n_sweeps=32)
model = nk.models.DeepSetRelDistance(hilbert=hilb,cusp_exponent=5,layers_phi=2,layers_rho=3,features_phi=(16, 16),features_rho=(16, 
16, 1))
vs = nk.vqs.MCState(sab, model, n_samples=4096, n_discard_per_chain=128)

with open("test.mpack", 'wb') as file:
  file.write(flax.serialization.to_bytes(vs))

op = nk.optimizer.Sgd(0.001)
sr = nk.optimizer.SR(diag_shift=0.01)

gs = nk.VMC(ha, op, sab, variational_state=vs, preconditioner=sr)
gs.run(n_iter=10, out="Helium_10_1d")
```
and there is the code g(r) when I reload the variational state from "test.mpack" :

```python
## load##
with open("test.mpack", 'rb') as file:
  vs = flax.serialization.from_bytes(vs, file.read())

hist_bins = 50
hist_range = (0, L / 2.0)

all_histograms = []
for sample in vs.samples:
    d = minimum_distance(sample, 1)
    hist, _ = np.histogram(d, bins=hist_bins, range=hist_range, density=True)
    all_histograms.append(hist)

         ## Computing average histogram##
average_hist = np.mean(all_histograms, axis=0)
bin_edges = np.linspace(hist_range[0], hist_range[1], hist_bins+1)

    ## Plotting histograms for all samples##
for hist in all_histograms:
    plt.plot(bin_edges[:-1], hist, alpha=0.3)

     ## Plotting the average histogram##
plt.plot(bin_edges[:-1], average_hist, 'r', linewidth=2)

plt.xlabel('Distance')
plt.ylabel('Probability Density')
plt.title('Histogram of Distances')
plt.show()
```
and this is the plot from the above code:

![Aziz-with-loading](https://github.com/netket/netket/assets/54255011/45dff7c3-f13a-4514-902d-3239eb06f84d)

but this is the plot of g(r) directly and without reloading "test.mpack":

![Aziz-without-loading](https://github.com/netket/netket/assets/54255011/2e521cb6-9a6c-4f29-94fd-b936aefa5b87)

Additionally in "https://netket.readthedocs.io/en/latest/docs/varstate.html" wrote when we want to reload the whole variational state we should use this code:
`
with open("test.mpack", 'rb') as file:
  vs.variables = flax.serialization.from_bytes(vs.variables, file.read())
`

but when I use it in reloading code of g(r) that you can see above, I get this error:

"The target dict keys and state dict keys do not match, target dict contains keys {'params'} which are not present in state dict at path ."
Thank you for your assistance.

**Comment by PhilipVinc** [comment, marked as answer]
There are two issues (I think?)
In the snippet above you are saving the Variational State *before* the optimisation. 
You should serialise it *after* the successful optimisation, so something more like

```python
gs = nk.VMC(ha, op, sab, variational_state=vs, preconditioner=sr)
gs.run(n_iter=1000, out="Helium_10_1d")

with open("test.mpack", 'wb') as file:
  file.write(flax.serialization.to_bytes(vs))
```

If you do this, you should get bit-perfect matching plots because the sampler will be synchronised and use the same random seed between before/after loading.

> Additionally in "https://netket.readthedocs.io/en/latest/docs/varstate.html" wrote when we want to reload the whole variational state we should use this code:
with open("test.mpack", 'rb') as file: vs.variables = flax.serialization.from_bytes(vs.variables, file.read())


Yes that is a small typo: if you serialise the full Variational State (`vs`) you should deserialise the full Variational State.
So...
```python
with open("vstate.mpack", 'wb') as file:
   file.write(flax.serialization.to_bytes(vs))
with open("vstate.mpack", 'rb') as file: 
   vs = flax.serialization.from_bytes(vs, file.read())
```

if you serialise only the variables you should deserialise only the variables
```python
with open("vars.mpack", 'wb') as file:
   file.write(flax.serialization.to_bytes(vs.variables))
with open("vars.mpack", 'rb') as file: 
   vars = flax.serialization.from_bytes(vs.variables, file.read())
vs.variables=vars
```

The first is serialising everything in the variational state, most notably the sampler and its random state and samples as well, while the latter is only serialising the variables or variational parameters, so if you reload it you will most likely end up with a different state for the sampler.

**Reply by MrAAafa99** [reply]
Thank you @PhilipVinc. This is exactly what I was looking for.
